src/balloonpanel/balloonpanelview.js

```diff
@@ -0,0 +1,371 @@
+ * @extends ui.View
+ */
+export default class BalloonPanelView extends View {
+	/**
+	 * @inheritDoc
+	 */
+	constructor( locale ) {
+		super( locale );
+
+		const bind = this.bind;
+
+		this.template = new Template( {
+			tag: 'div',
+			attributes: {
+				class: [
+					'ck-balloon-panel',
+					'ck-link-balloon-panel',
+					bind.to( 'arrow', ( value ) => `ck-balloon-panel_arrow_${ value }` ),
+					bind.if( 'isVisible', 'ck-balloon-panel_visible' )
+				],
+
+				style: {
+					top: bind.to( 'top', ( value ) => `${ value }px` ),
+					left: bind.to( 'left', ( value ) => `${ value }px` ),
+					maxWidth: bind.to( 'maxWidth', ( value ) => `${ value }px` )
+				}
+			}
+		} );
+
+		this.register( 'content', el => el );
+
+		/**
+		 * Model of this balloon panel view.
+		 *
+		 * @member {ui.balloonPanel.BalloonPanelView} ui.balloonPanel.BalloonPanelView#model
+		 */
+	}
+
+	/**
+	 * Check if balloon is currently visible.
+	 *
+	 * @returns {Boolean}
+	 */
+	get isVisible() {
+		return this.model.isVisible;
+	}
+
+	/**
+	 * Show balloon and start listen to `keydown` event for closing by `Esc` key.
+	 */
+	show() {
+		this.model.isVisible = true;
+
+		// Attach keydown listener for closing panel on Esc press.
+		this._keyCloseCallback = evt => {
+			if ( evt.keyCode == 27 ) {
+				this.hide();
+			}
+		};
+		document.addEventListener( 'keydown', this._keyCloseCallback );
+	}
+
+	/**
+	 * Hide balloon, and stop listen `keydown` event.
+	 *
+	 * @fires ui.balloonPanel.BalloonPanelModel#hide
+	 */
+	hide() {
+		this.model.isVisible = false;
+		this.model.fire( 'hide' );
+		document.removeEventListener( 'keydown', this._keyCloseCallback );
+	}
+
+	/**
+	 * Attach balloon panel to specified element or range including limits of viewport bounding box.
+	 *
+	 *  Balloon could show in 4 relative to the target element places:
+	 * * South east:
+	 *
+	 * 		[ Target ]
+	 *		    ^
+	 *		+-----------------+
+	 *		|                 |
+	 *		+-----------------+
+	 *
+	 *
+	 * * South west:
+	 *
+	 * 		         [ Target ]
+	 *		              ^
+	 *		+-----------------+
+	 *		|                 |
+	 *		+-----------------+
+	 *
+	 *
+	 * * North east:
+	 *
+	 * 		+-----------------+
+	 * 		|                 |
+	 *		+-----------------+
+	 *		    V
+	 *		[ Target ]
+	 *
+	 *
+	 * * North west:
+	 *
+	 * 		+-----------------+
+	 * 		|                 |
+	 *		+-----------------+
+	 *		              V
+	 *		         [ Target ]
+	 *
+	 * @param {HTMLElement|Range} elementOrRange Target element to which the balloon will be attached.
+	 * @param {HTMLElement|Object} limiter Viewport element.
+	 */
+	attachTo( elementOrRange, limiter ) {
+		this.show();
+
+		const elementOrRangeRect = new AbsoluteDomRect( elementOrRange, 'elementOrRange' );
+		const panelRect = new AbsoluteDomRect( this.element );
+		const visibleInViewportRect = getVisibleInViewportRect( limiter );
+
+		// Get bounding box of each available place and move balloon in this place where fits the best.
```

Comment:
"Create the rect for each of the possible balloon positions and use the one which is the best."